### PR TITLE
Add method `eq_ignore_case` to `str`, `char`, and `Chars`

### LINKED ISSUE #57221: Add `eq_ignore_case` for `str` and `char`.
In `str` and `char`, there are functions `eq_ignore_ascii_case`.  We should also implement `eq_ignore_case` for these types.

### PR DESCRIPTION
**Pull request by czipperz**
Resolves #57221

### CODE REVIEW

**Comment by czipperz** [comment]
I might note I have no little to no clue what I'm doing.  I based the `unstable` annotation upon some other code I saw in the files.  To do: add feature `eq_ignore_case`.  Not sure how to do that.

**Comment by czipperz** [comment]
@the8472 thanks for the help will do

**Comment by BurntSushi** [comment]
My guess is that if caseless matching isn't available in an easily spottable way, then folks will continue to use `to_lowercase`/`to_uppercase` to do caseless matching. I don't know how common it is in general, but a key advantage of `to_lowercase`/`to_uppercase` is that they are more flexible than the APIs provided by `caseless` in that they allow one to re-arrange comparisons by making different trade offs. e.g., storing a list of case "normalized" stop words in a set and using that to determine whether any given word is considered a stop word. You can't do that with `caseless`. I suspect that's why crates like [`unicase`](https://docs.rs/unicase) have appeared, although unicase only seems to support default caseless matching.

In practice, using `to_lowercase` works quite well until you hit input where it doesn't, and it's unlikely that those types of inputs will be unit tested unless the programmer is intimately familiar with Unicode.

For that reason, I could personally see a justification for these types of routines entering std, but I think before we even consider that, we need a more prominent crate in the ecosystem that correctly and simply handles most of the use cases that folks care about without them having to learn how to read the Unicode Standard. Otherwise, folks are (understandably) going to continue doing the wrong thing.

In summary, IMO, this stuff should be prototyped in a high quality ecosystem crate first. I don't know how to determine whether it should go into std from there, but if I had perfect knowledge, it would be something like, "when people need caseless matching, do they use crate X to do it, or do they do it incorrectly with `to_lowercase` because that's what std gave them and they didn't know any better?" If the latter, and if we believe moving caseless matching into std would help fix it, then maybe we should take that path.

**Comment by SimonSapin** [comment]
> they allow one to re-arrange comparisons by making different trade offs. e.g., storing a list of case "normalized" stop words in a set and using that to determine whether any given word is considered a stop word. You can't do that with caseless.

What does "stop word" mean in this context?

The `caseless` carte does provide `default_case_fold`. You could keep around the result of `string.chars().default_case_fold().collect::<String>()` and compare it later, as a different way to implement `default_caseless_match`.

(But yes, lack of docs makes this hard to discover.)

----

I hear the argument that people will keep doing the wrong thing if that’s all that is available. Maybe we should add something here in std, but it’s not at all obvious to me what.

Possibly default case folding and default caseless matching. They have "default" in their names. But the very next sentence in Unicode after defining default caseless matching is:

> When  comparing  strings  for  case-insensitive  equality,  the  strings  should  also  be  normalized for most correct results.

… which introduces canonical caseless matching. In other words, default is often incorrect and canonical should (maybe? probably?) be preferred.

We can add canonical caseless matching to std. This involves shipping an implementation of NFD normalization. Since we’re shipping it anyway, should we expose it in public a API? If we’re doing that, should we add other normalization forms as well? All of the `unicode-normalization` crate’s API?

**Comment by BurntSushi** [comment]
> What does "stop word" mean in this context?

It's just an example use case to demonstrate one might want to trade space for time. I think the generic description of a stop word is suitable here: https://en.wikipedia.org/wiki/Stop_words

With that said, you're right, `caseless` does meet that use case. I missed the `default_case_fold` iterator. :-)

> … which introduces canonical caseless matching. In other words, default is often incorrect and canonical should (maybe? probably?) be preferred.

> We can add canonical caseless matching to std. This involves shipping an implementation of NFD normalization. Since we’re shipping it anyway, should we expose it in public a API? If we’re doing that, should we add other normalization forms as well? All of the unicode-normalization crate’s API?

Sure, I appreciate that, which is why I think we should start by building a high quality crate in the ecosystem first, right? Like, are we really trying to design out the full API, correctness trade offs and implementation details right in this PR review? I mean, Unicode's algorithms are chalk full of correctness trade offs. Many of their algorithms have a variety of tailorings that are rarely implemented in my experience. Does this mean we should just not do anything? If shipping normalization is a problem but default caseless matching without normalization improves the status quo, then yeah, maybe that's good to have in std? I'm just stating a path as I see it, but the path does not begin in std.

**Comment by SimonSapin** [comment]
@BurntSushi I basically agree with you on everything here. *Something* should be probably added to std to improved the status quo but this is full of correctness trade offs, and needs more design work than we’re gonna do in PR comments.

Unfortunately this likely won’t make much progress until someone champions a design.
